### DISCUSSION TITLE
No .f90 in hemi and turbChannel examples

### DISCUSSION THREAD

**Discussion by timofeymukha** [ExtremeFLOW/neko#338] [General]
Basically, subj.
For hemi there is one here
https://github.com/ExtremeFLOW/test-suite/blob/master/neko_hemi
can it be reused?

**Comment by njansson** [comment]
The case in `src/examples/hemi` is an example of "standalone" case whereas inflow etc is all defined in the core solver.

To run it simply copy the neko binary from the `src` (or from `bin` under the installation prefix) and run with `neko hemi.case`

However, the current version of this case is slightly different from the test suite. As default it uses a uniform inflow profile, but it could be changed by adding/modifying the following in the case file,


```
&NEKO_CASE
...
initial_condition = 'blasius'
/
```

and 

```
&NEKO_PARAMETERS
...
fluid_inflow = 'blasius'
/
```

there's also a parameter `delta` and `blasius_approx` under `NEKO_PARAMETERS` related to the boundary layer thickness (default 1) and the profile approximation (default sin).

**Comment by timofeymukha** [comment]
Ah, I see. Nice to be able to run "standalone". I guess with modern fortran we can go quite far in terms of allowing to specify everything in the config instead of tinkering in user-defined functions.

**Comment by njansson** [comment]
Correct, that is at least the intent. Of course one can't cover all the possibilities, but it will make life much easier for users on accelerators if "standalone" would cover most of their cases (otherwise people would need to write accelerator code in their user file)

**Comment by timofeymukha** [comment]
@njansson why does the channel case not use `avflow`, `vol_flow_dir`, `flow_rate` to drive the flow? On the other hand, `uinf` is set, what does that do exactly?

**Reply by njansson** [reply]
Definitely due to neglect from our side, the case is not setup correct afaik. @MartinKarp can correct me but I don't think we had `vol_flow_dir` implemented when we use that case to debug periodic conditions.

Ps. `vol_flow` and friends are currently only available in the `plan4` solver. The pipe case is a good example on how to use it

**Reply by MartinKarp** [reply]
Yeah, a forcing or volflow is necessary for the channel case.

**Reply by njansson** [reply]
Patches are more than welcomed  (^-^)

**Reply by timofeymukha** [reply]
![channel](https://user-images.githubusercontent.com/7013791/171494971-90c66d6c-2b3f-4ccb-93f8-d0e1cd1f1c3a.PNG)

**Reply by njansson** [reply]
`vol_flow` should now work with the `fluid_pnpn` solver. A device version will appear soon, progress can be tracked in #554

**Comment by Bkv-1915** [comment]
HI. I am trying to run the turb_channel example on my GPU. However, it says that it will run in 10 minutes. I have RTX 4090 in my workstation and it takes 150 minutes. Am I missing something?

**Reply by timfelle** [reply]
Which command did you use to setup Neko and which command did you use to execute the example? 
This sounds to me like you are either not using the gpu or you are launching more than one mpi ranks which will then compete for using your gpu.

**Reply by Bkv-1915** [reply]
I used : ./configure  --with-cuda=/usr/local/cuda --prefix=/path to setup Neko. To execute the example, I used: mpirun -np 24 ./neko turb_channel.case. I did this after /path/bin/makeneko turb_channel.f90. FYI I am new to this. I am sorry if there is a basic mistake here. 

edit: I checked the processes and it says 100% GPU and 99% CPU is being used while the simulation is runnig.

edit2: Its way quicker when I do mpirun -np 1 ./neko turb_channel.case. Thanks @timfelle

**Reply by timfelle** [reply]
Great, I am happy that helped resolve the issue. I will also just take a look at our documentation and see if we actually are clear enough about that little detail 😅 It might be one of those that have slipped through the cracks.
